### PR TITLE
Change of index values dtype should be reflected in index type

### PR DESCRIPTION
**Pull request by josteinbf**
This PR should eventually address issue #9966. I will look at testing first to make sure I have understood how pandas ought to work in these situations.

### CODE REVIEW

**Comment by josteinbf** [comment]
Tests should fail for now.

@jreback Do you agree that the inplace multiplication in 6279449 should raise an Exception? I guess what I'm really asking is if there is a way an object can change its type in response to an inplace operation? There might be, because this works for simple ints and floats.

**Comment by jreback** [comment]
in place are disable as Index are disabled anyhow
so u can test for this but it shouldn't work (eg test that it raises)

File: pandas/tests/test_index.py [hunk 1]
```diff
@@ -1909,6 +1909,46 @@ def test_astype(self):
...
+                     operator.floordiv)
+        for op in operators:
+            index_promoted = op(index_int, np.pi)
+            self.assertEqual(index_promoted.dtype, np.float64)
+            self.assertIsInstance(index_promoted, Float64Index)
+
+    def test_promote_type_inplace(self):
+        # Related to GH-9966
+        index = Int64Index(np.arange(5))
+        index += np.pi
```
Comment (jreback) [inline pandas/tests/test_index.py]: these should all fail with a TypeError (IIRC). inplace ops are not allowed on an immutable type (I guess its not actually disabled though ).

Comment (shoyer) [inline pandas/tests/test_index.py]: I know that in some cases Python will automatically treat an in-place operation like `x += y` as `x = x + y`, e.g., if you implement `__add__` but not `__iadd__`. This is how inplace operations work on integers, for example, which are immutable.

In any case, this should certainly fail with `TypeError` if `index` is the same object on both these lines.

**Comment by jreback** [comment]
want to update for 0.17.0?

**Comment by jreback** [comment]
can you update?

**Comment by josteinbf** [comment]
Sorry I haven't gotten back to you before. I'm afraid I'll be to busy to be
able to finish the remaining parts here.

2015-06-27 1:34 GMT+02:00 jreback <notifications@github.com>:

**Comment by jreback** [comment]
closing in favor of #10638

**Pull request closed without merging**